endpoint, endpointmanager: Publish max policymap size as metric

christarazi opened the pull request:
This commit reworks the cilium_bpf_map_pressure metric such that the max
policymap size from all endpoints is published. This reduces the
cardinality of the metric because each individual endpoint policymap was
a unique label on the metric. In other words, this commit collapses the
cardinality down to just one.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

---

Tasks:

- [x] Manual testing
- [x] Update docs
- [x] Update release note
- [x] Remove unnecessary logs
- [x] Document why 10s for trigger interval
- [x] Decide on label name (i.e. `cilium_policy_*`?)

Fixes: https://github.com/cilium/cilium/issues/27332

```
`cilium_bpf_map_pressure` now has reduced cardinality for endpoint policymaps. Only one label is exposed representing the maximum policymap pressure on the node.
```

christarazi committed:
endpoint, endpointmanager: Publish max policymap size as metric

This commit reworks the cilium_bpf_map_pressure metric such that the max
policymap size from all endpoints is published. This reduces the
cardinality of the metric because each individual endpoint policymap was
a unique label on the metric. In other words, this commit collapses the
cardinality down to just one.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/endpointmanager/policymap_pressure.go:
```diff
@@ -0,0 +1,66 @@
+
+	var err error
+	src, emit, _ := stream.Multicast[endpoint.MapPressureEvent]() // ignoring complete() here as there's no single emitter that can complete it
+	mgr.policyMapPressureEventCh = stream.ToChannel(mgr.ctx, src)
```
Buffered channel?

christarazi replied:
Removed channel and usage of `pkg/stream`

christarazi commented on pkg/endpointmanager/policymap_pressure.go:
```diff
@@ -0,0 +1,66 @@
+			log.Infof("EndpointManager policymap event loop terminated")
+			return
+		}
+	}
```
If we use buffered channels, then the loop should be written as 

```
for range mgr.policyMapPressureEventCh
```

and when it's closed, the loop terminates. If the context is cancelled, we should drain the channel as well.

However, this is starting to sound like pkg/eventqueue. :thinking:

christarazi replied:
Removed channel and usage of `pkg/stream`

joamaki commented on pkg/endpointmanager/policymap_pressure.go:
```diff
@@ -0,0 +1,65 @@
+
+	var err error
+	src, emit, _ := stream.Multicast[endpoint.MapPressureEvent]() // ignoring complete() here as there's no single emitter that can complete it
+	mgr.policyMapPressureEventCh = stream.ToChannel(mgr.ctx, src)
```
I would just do `src.Observe(mgr.ctx, mgr.onPolicyMapPressureEvent, func(error) {})` here instead of going via a channel. Then one doesn't need to worry about draining the channel etc. concerns.

Since there's only once consumer for the `MapPressureEvent`, do we actually benefit at all from the observable? Essentially "Multicast" gives you a many-to-many "callback". If you've got many-to-one, I'd probably just call a method, e.g. `EndpointManager.UpdatePolicyMapPressure(...)` (or `PolicyMapPressure.Update` implemented by EndpointManager if cyclic imports is a concern).

christarazi replied:
Removed channel and usage of `pkg/stream`. Rewrote it such that the call to update the metric is behind an interface that the endpointmanager implements.

joamaki commented on pkg/endpointmanager/policymap_pressure.go:
```diff
@@ -0,0 +1,65 @@
+				mgr.policyMapPressureTrigger.Trigger()
+			}
+		case <-mgr.ctx.Done():
+			log.Infof("EndpointManager policymap event loop terminated")
```
This won't drain the channel and will block the emitter.

christarazi replied:
Removed channel and usage of `pkg/stream`

christarazi committed:
docs: Update docs re: cilium_bpf_map_pressure for policy maps

Now update the metrics reference and the release note after the previous
commit consolidated the cardinality of the metric from per policymap to
just one label.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

qmonnet commented on pkg/endpointmanager/policymap_pressure.go:
```diff
@@ -0,0 +1,81 @@
+	log.WithField(logfields.Value, val).Debug("EndpointManager policymap received event")
+
+	cur := p.current.Load()
+	// TODO: Do we need to define a threshold for the comparison?
```
Does this need to be addressed before merging?

christarazi replied:
Dropped it. I think it's has minimal consequence anyway.

qmonnet reviewed (approved):
Looks good from my side, thanks!

aditighag commented on pkg/endpointmanager/policymap_pressure.go:
```diff
@@ -0,0 +1,80 @@
+		Name:        "endpointmanager-policymap-max-size-metrics",
+	})
+	if err != nil {
+		log.WithError(err).Panic("Failed to initialize trigger for policymap pressure metric")
```
It's not common to panic in the agent. Why is this an exception?

christarazi replied:
In this case, the only way this panic gets triggered is if the `TriggerFunc` is nil. Given this code will initialize very early on in the daemon, it acts as a sanity check before proceeding. It is assumed that if this code panics, then it was a developer mistake. There is no risk that this panics in production because it is only called once on dameon init and it will be caught immediately in CI. Furthermore, the endpointmanager is not created conditionally nor is that likely to change in the future; in other words, this init code will always execute meaning if the `TriggerFunc` is ever nil, it will be caught.

joestringer replied:
I feel like we could probably drop the error condition from `NewTrigger()` and just force the caller to pass a function directly in, to ensure it isn't nil. It doesn't really make sense for a caller to initialize a trigger without a function to trigger, since that defeats the whole point of the trigger in the first place. That could help to avoid the need for this sort of awkward error checking + panic logic in arbitrary packages like this.

(I buy your argument in this specific scenario, though I am a bit paranoid in general about `Panic()`, specifically because of the potential for unexpected agent failure)

joestringer replied:
^^ Doesn't need to be addressed in this PR, just a suggestion for future.

aditighag reviewed (commented):
>The ``cilium_bpf_map_pressure`` for policy maps is now exposed as a single
  label ``cilium_policy_*``, rather than a label per policy map of an endpoint.

There is some context missing in the PR. Why is an overall pressure metric preferred over a per endpoint metric? As a user, it would be beneficial to know the top-k endpoint policy maps under pressure. What if there are multiple endpoint policy maps experiencing similar pressure?
It would also be useful to add a sample snippet for the updated map pressure metric.

aditighag reviewed (approved):
Thanks for the summary. It'll be great to see the troubleshooting examples as follow-ups.
